# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
import math
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error
import scipy.stats as st

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
x.shape

(506, 13)

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

<b> Write code for generating samples</b><

In [4]:
def generating_samples(input_data, target_data):
    '''In this function, we will write code for generating 30 samples '''

    sel_rows = np.random.choice(range(506),303,replace = False)
    repl_rows = np.random.choice(range(303),203,replace = False)
    sel_cols = np.random.choice(range(13),10,replace = False)

    sample_data = input_data[sel_rows[:, None], sel_cols]
    target_sample = target_data[sel_rows]


    # Replicate Rows
    repl_sample_data = sample_data[repl_rows]
    repl_target_sample = target_sample[repl_rows]

    # Concat
    final_sample = np.vstack((sample_data,repl_sample_data))
    final_target_sample = np.vstack((target_sample.reshape(-1,1),repl_target_sample.reshape(-1,1)))
    

    return final_sample , final_target_sample,sel_rows,sel_cols

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [5]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)


True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [27]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
  a,b,c,d = generating_samples(x, y)
  list_input_data.append(a)
  list_output_data.append(b)
  list_selected_row.append(c)
  list_selected_columns.append(d)


<font color='cyan'> <b>Grader function - 2 </b></font>

In [28]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [29]:
model_list = []

for idx in range(30):
  model = DecisionTreeRegressor(max_depth = None)
  model.fit(list_input_data[idx], list_output_data[idx])
  model_list.append(model)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <b> Write code for calculating MSE</b>

In [30]:
def cal_MSE(x,models,list_selected_row_,list_selected_columns_):
  y_preds = []
  for x_pt in x:
    temp_y = []
    for i in range(30):
      # data_point = x[i]
      sel_cols = list_selected_columns_[i]
      res = models[i].predict(x_pt[sel_cols].reshape(1,-1))
      temp_y.append(res)

    y_preds.append(np.median(temp_y))
  
  return y_preds


In [31]:
y_preds = cal_MSE(x,model_list,list_selected_row,list_selected_columns)
MSE = mean_squared_error(y,y_preds)
print("MSE : ",MSE)

MSE :  0.03768280632411066


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <b> Write code for calculating OOB score </b>

In [32]:
def cal_OOB(x,models,list_selected_row_,list_selected_columns_):
  y_preds_oob = []
  for x_idx,x_pt in enumerate(x):
    temp_y = []
    for i in range(30):
      sel_rows = list_selected_row_[i]
      # if data_point not in dataset : pass 
      if x_idx in sel_rows:
        continue

      sel_cols = list_selected_columns_[i]
      res = models[i].predict(x_pt[sel_cols].reshape(1,-1))
      temp_y.append(res)

    y_preds_oob.append(np.median(temp_y))

  return y_preds_oob


In [33]:
y_preds_oob = cal_OOB(x,model_list,list_selected_row,list_selected_columns)
OOB = mean_squared_error(y,y_preds_oob)
print("OOB Score : ",OOB)

OOB Score :  13.227707509881423


In [37]:
predictions = []
for i in range(len(x)):
  data = x[i]
  preds = []
  for j in range(30):
    train_data_rows = list_selected_row[j]
    train_data_cols = list_selected_columns[j]

    if i not in train_data_rows:
      temp_res = model_list[j].predict(data[train_data_cols].reshape(1,-1))
      preds.append(temp_res)

  predictions.append(np.median(preds))

In [38]:
print(mean_squared_error(y,predictions))

13.227707509881423


# <font color='blue'><b>Task 2</b></font>

In [ ]:
from tqdm import tqdm

In [ ]:
def run_task_1(n):
  mse_list = []
  oob_list = []
  for i in tqdm(range(n)):
    # Use generating_samples function to create 30 samples 
    list_input_data_ =[]
    list_output_data_ =[]
    list_selected_row_= []
    list_selected_columns_=[]
    for i in range(0,30):
      a_,b_,c_,d_ = generating_samples(x, y)
      list_input_data_.append(a_)
      list_output_data_.append(b_)
      list_selected_row_.append(c_)
      list_selected_columns_.append(d_)


    #Building Decision Trees
    model_list_new = []
    for idx in range(30):
      model = DecisionTreeRegressor(max_depth = None)
      model.fit(list_input_data_[idx], list_output_data_[idx])
      model_list_new.append(model)

    #MSE & OOB
    mse = mean_squared_error(y,cal_MSE(x,model_list_new,list_selected_row_,list_selected_columns_))
    oob = mean_squared_error(y,cal_OOB(x,model_list_new,list_selected_row_,list_selected_columns_))

    mse_list.append(mse)
    oob_list.append(oob)

  return mse_list,oob_list
  


In [ ]:
mse_list,oob_list = run_task_1(30)

100%|██████████| 30/30 [00:43<00:00,  1.44s/it]


${\text{Upper 95%  limit}{\displaystyle ={\bar {x}}+{\text{SE}}\times 1.96}}$ <br>
${\text{Lower 95%  limit}{\displaystyle ={\bar {x}}-{\text{SE}}\times 1.96}}$


In [ ]:
# Z Table using scipy : https://stackoverflow.com/questions/20864847/probability-to-z-score-and-vice-versa
def cal_CI(sample,ci_percent):
  x_mean = np.mean(sample)
  x_std = np.std(sample)
  SE = x_std/np.sqrt(len(sample))

  z_score = st.norm.ppf(ci_percent/100)

  upper_lim = x_mean + (z_score*SE)
  lower_lim = x_mean - (z_score*SE)

  return lower_lim,upper_lim

In [ ]:
from prettytable import PrettyTable
table = PrettyTable()
table.field_names = ["METRIC", "CONFIDENCE PERCENTAGE", "LOWER","VALUE"]

hp = [98,97,96,95,90,80,70]
for p in hp:
  low, up = cal_CI(mse_list,p)
  table.add_row(["MSE", str(p)+"%", np.round(low,6), np.round(up,6)])
for p in hp:
  low, up = cal_CI(oob_list,p)
  table.add_row(["OOB", str(p)+"%", np.round(low,6), np.round(up,6)])

print(table)


+--------+-----------------------+-----------+-----------+
| METRIC | CONFIDENCE PERCENTAGE |   LOWER   |   VALUE   |
+--------+-----------------------+-----------+-----------+
|  MSE   |          98%          |  0.082337 |  0.144839 |
|  MSE   |          97%          |  0.084969 |  0.142207 |
|  MSE   |          96%          |  0.086949 |  0.140228 |
|  MSE   |          95%          |  0.088559 |  0.138617 |
|  MSE   |          90%          |  0.094087 |  0.133089 |
|  MSE   |          80%          |  0.100782 |  0.126395 |
|  MSE   |          70%          |  0.105609 |  0.121568 |
|  OOB   |          98%          | 12.170694 | 13.098279 |
|  OOB   |          97%          | 12.209752 | 13.059221 |
|  OOB   |          96%          | 12.239134 | 13.029839 |
|  OOB   |          95%          | 12.263034 | 13.005939 |
|  OOB   |          90%          | 12.345078 | 12.923896 |
|  OOB   |          80%          | 12.444426 | 12.824548 |
|  OOB   |          70%          | 12.516063 | 12.752911

# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

In [ ]:
xq= np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])

y_pred = []
for i in range(30):
  sel_cols = list_selected_columns[i]
  pred = model_list[i].predict(xq[sel_cols].reshape(1,-1))
  y_pred.append(pred)

print("Predicted House Price : ",np.median(y_pred))


Predicted House Price :  18.7


<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

**TASK 1**
- The observed MSE was 0.037028 and OOB score was 12.2415
- On repeated execution, the MSE was found to vary between 0.01 to 0.2 and OOB score was usually between 10-30
- If number of samples are increased then it would lead to better result and lower MSE

**TASK 2** <br>
- As we dont have population standard deviation, we assume mean sd = population sd
- Mean MSE Score between 0.088559 and 0.138617 with 95% confidence
- Mean OOB Score lies between 12.263034 and 13.005939 with 95% confidence

**TASK 3**

- Predicted House Price for the given data point :  18.7

In [ ]:
!jupyter nbconvert --to html "/content/Bootstrap_assignment.ipynb"

[NbConvertApp] Converting notebook /content/Bootstrap_assignment.ipynb to html
[NbConvertApp] Writing 329575 bytes to /content/Bootstrap_assignment.html
